lib/utils/compose.coffee

```diff
@@ -0,0 +1,747 @@
+			docker.getImage(image.name).inspect().get('Size').then (size) ->
+				image.props.size = size
+			.return(image)
+		.tap (images) ->
+			summary = _(images).map ({ serviceName, props }) ->
+				[ serviceName, "Image size: #{humanize.filesize(props.size)}" ]
+			.fromPairs()
+			.value()
+			renderer.end(summary)
+	.finally(renderer.end)
+
+createRelease = (apiEndpoint, auth, userId, appId, composition) ->
+	_ = require('lodash')
+	crypto = require('crypto')
+	releaseMod = require('resin-release')
+
+	client = releaseMod.createClient({ apiEndpoint, auth })
+
+	releaseMod.create
+		client: client
+		user: userId
+		application: appId
+		composition: composition
+		source: 'local'
+		commit: crypto.pseudoRandomBytes(16).toString('hex').toLowerCase()
+	.then ({ release, serviceImages }) ->
+		release = _.omit(release, [
+			'created_at'
+			'belongs_to__application'
+			'is_created_by__user'
+			'__metadata'
+		])
+		_.keys serviceImages, (serviceName) ->
+			serviceImages[serviceName] = _.omit(serviceImages[serviceName], [
+				'created_at'
+				'is_a_build_of__service'
+				'__metadata'
+			])
+		return { client, release, serviceImages }
+
+tagServiceImages = (docker, images, serviceImages) ->
+	Promise.map images, (d) ->
+		serviceImage = serviceImages[d.serviceName]
+		imageName = serviceImage.is_stored_at__image_location
+		[ _match, registry, repo, tag = 'latest' ] = /(.*?)\/(.*?)(?::([^/]*))?$/.exec(imageName)
+		name = "#{registry}/#{repo}"
+		docker.getImage(d.name).tag({ repo: name, tag, force: true })
+		.then ->
+			docker.getImage("#{name}:#{tag}")
+		.then (localImage) ->
+			serviceName: d.serviceName
+			serviceImage: serviceImage
+			localImage: localImage
+			registry: registry
+			repo: repo
+			logs: d.logs
+			props: d.props
+
+authorizePush = (tokenAuthEndpoint, registry, images) ->
+	_ = require('lodash')
+	sdk = require('resin-sdk').fromSharedOptions()
+
+	if not _.isArray(images)
+		images = [ images ]
+
+	sdk.request.send
+		baseUrl: tokenAuthEndpoint
+		url: '/auth/v1/token'
+		qs:
+			service: registry
+			scope: images.map (repo) ->
+				"repository:#{repo}:pull,push"
+	.get('body')
+	.get('token')
+	.catchReturn({})
+
+pushAndUpdateServiceImages = (docker, token, images, afterEach) ->
+	chalk = require('chalk')
+	{ DockerProgress } = require('docker-progress')
+	tty = require('./tty')(process.stdout)
+
+	opts = { authconfig: registrytoken: token }
+
+	progress = new DockerProgress(dockerToolbelt: docker)
+	renderer = pushProgressRenderer(tty, chalk.blue('[Push]') + '    ')
+	reporters = progress.aggregateProgress(images.length, renderer)
+
+	Promise.using tty.cursorHidden(), ->
+		Promise.map images, ({ serviceImage, localImage, props, logs }, index) ->
+			Promise.join(
+				localImage.inspect().get('Size')
+				progress.push(localImage.name, reporters[index], opts).finally(renderer.end)
+				(size, digest) ->
+					serviceImage.image_size = size
+					serviceImage.content_hash = digest
```

Comment:
If I repeatedly rebuild the same thing, with the entire build coming from the cache, the build is super fast but I still end up doing the entire image upload from scratch. I've checked the API though, and I do end up with the same digest for the image, so in principle we should be able to detect that.

This doesn't need to be in this PR, but as a future improvement is there an easy way we could avoid the image upload entirely in those cases?

Reply:
Yes, that's true unfortunately. I did try keeping images around between consecutive deploys, but that didn't help. The issue is either that docker doesn't keep around the metadata needed to determine that layers have been pushed already (which means simply pulling the previous images will fix the issue) or there's something with the way the API authenticates access to the registry. Didn't get round to test either theory, but I suspect it should be straight-forward.